src/core/cryptor.rs

```diff
@@ -1,59 +1,98 @@
+};
 
-/// This trait is used to encrypt and decrypt messages sent to the
-/// [`PubNub API`].
+/// Encrypted data representation object.
 ///
-/// It is used by the [`dx`] modules to encrypt messages sent to PubNub and
-/// returned by the [`PubNub API`].
+/// Objects contain both encrypted data and additional data created by cryptor
+/// that will be required to decrypt the data.
+#[derive(Debug)]
+pub struct EncryptedData {
+    /// Cryptor-defined information.
+    ///
+    /// Cryptor may provide here any information which will be useful when data
+    /// should be decrypted.
+    ///
+    /// For example `metadata` may contain:
+    /// * initialization vector
+    /// * cipher key identifier
+    /// * encrypted `data` length.
+    pub metadata: Option<Vec<u8>>,
+
+    /// Encrypted data.
+    pub data: Vec<u8>,
+}
+
+/// Cryptor trait.
+///
+/// Types that implement this trait can be used to configure [`CryptoProvider`]
+/// implementations for standalone usage or as part of [`PubNubClientInstance`]
+/// for automated data _encryption_ and _decryption_.
 ///
 /// To implement this trait, you must provide `encrypt` and `decrypt` methods
-/// that takes a `&[u8]` and returns a `Result<Vec<u8>, PubNubError>`.
+/// that takes a `Vec<u8>` and returns a `Result<EncryptedData, PubNubError>`.
 ///
 /// You can implement this trait for your own types, or use one of the provided
-/// features to use a crypto library.
-/// When you use this trait to make your own crypto, make sure that other SDKs
-/// use the same encryption and decryption algorithms.
+/// features to use a `crypto` library.
+///
+/// You can implement this trait for your cryptor types, or use one of the
+/// implementations provided by `crypto` feature.
+/// When you implement your cryptor for custom encryption and use multiple
+/// platforms, make sure that the same logic is implemented for other SDKs.
 ///
 /// # Examples
 /// ```
-/// use pubnub::core::{Cryptor, error::PubNubError};
+/// use pubnub::core::{Cryptor, EncryptedData, error::PubNubError};
 ///
 /// #[derive(Debug)]
 /// struct MyCryptor;
 ///
 /// impl Cryptor for MyCryptor {
-///     fn encrypt(&self, source: Vec<u8>) -> Result<Vec<u8>, PubNubError> {
+///     fn identifier(&self) -> [u8; 4] {
+///         *b"MCID"
+///     }
+///
+///     fn encrypt(&self, source: Vec<u8>) -> Result<EncryptedData, PubNubError> {
 ///         // Encrypt provided data here
-///         Ok(vec![])
+///         Ok(EncryptedData {
+///             metadata: None,
+///             data: vec![]
+///         })
 ///     }
 ///
-///     fn decrypt(&self, source: Vec<u8>) -> Result<Vec<u8>, PubNubError> {
+///     fn decrypt(&self, source: EncryptedData) -> Result<Vec<u8>, PubNubError> {
 ///         // Decrypt provided data here
 ///         Ok(vec![])
 ///     }
 /// }
 /// ```
-///
-/// [`dx`]: ../dx/index.html
-/// [`PubNub API`]: https://www.pubnub.com/docs
 pub trait Cryptor: Debug + Send + Sync {
```

Comment:
as above

Reply:
@Xavrax because `Cryptor` is part of another trait / interface which derives from `Debug` it also needs to have `Debug`.